src/components/Launch/LaunchWorkflowForm/launchMachine.ts

```diff
@@ -0,0 +1,448 @@
+            }
+        },
+        [LaunchState.SELECT_TASK_VERSION]: {
+            // events handled at top level
+        },
+        [LaunchState.LOADING_INPUTS]: {
+            invoke: {
+                src: 'loadInputs',
+                onDone: {
+                    target: LaunchState.ENTER_INPUTS,
+                    actions: ['setInputs']
+                },
+                onError: {
+                    target: LaunchState.FAILED_LOADING_INPUTS,
+                    actions: ['setError']
+                }
+            }
+        },
+        [LaunchState.FAILED_LOADING_INPUTS]: {
+            on: {
+                RETRY: LaunchState.LOADING_INPUTS
+            }
+        },
+        [LaunchState.UNSUPPORTED_INPUTS]: {
+            // events handled at top level
+        },
+        [LaunchState.ENTER_INPUTS]: {
+            always: {
+                target: LaunchState.UNSUPPORTED_INPUTS,
+                cond: ({ unsupportedRequiredInputs }) =>
+                    unsupportedRequiredInputs.length > 0
+            },
+            on: {
+                SUBMIT: LaunchState.SUBMIT_VALIDATING,
+                VALIDATE: LaunchState.VALIDATING_INPUTS
+            }
+        },
+        [LaunchState.VALIDATING_INPUTS]: {
+            invoke: {
+                src: 'validate',
+                onDone: LaunchState.ENTER_INPUTS,
+                onError: LaunchState.INVALID_INPUTS
+            }
+        },
+        [LaunchState.INVALID_INPUTS]: {
+            on: {
+                VALIDATE: LaunchState.VALIDATING_INPUTS
+            }
+        },
+        [LaunchState.SUBMIT_VALIDATING]: {
+            invoke: {
+                src: 'validate',
+                onDone: {
+                    target: LaunchState.SUBMITTING
+                },
+                onError: {
+                    target: LaunchState.INVALID_INPUTS,
+                    actions: ['setError']
+                }
+            }
+        },
+        [LaunchState.SUBMITTING]: {
+            invoke: {
+                src: 'submit',
+                onDone: {
+                    target: LaunchState.SUBMIT_SUCCEEDED,
+                    actions: ['setExecutionId']
+                },
+                onError: {
+                    target: LaunchState.SUBMIT_FAILED,
+                    actions: ['setError']
+                }
+            }
+        },
+        [LaunchState.SUBMIT_FAILED]: {
+            on: {
+                SUBMIT: LaunchState.SUBMITTING
+            }
+        },
+        [LaunchState.SUBMIT_SUCCEEDED]: {
+            type: 'final'
+        }
+    }
+};
+
+/** A full machine for representing the Launch flow, combining the state definitions
+ * with actions/guards/services needed to support them.
+ */
+export const launchMachine = Machine(launchMachineConfig, {
+    actions: {
+        setWorkflowVersion: assign((_, event) => ({
+            workflowVersion: (event as SelectWorkflowVersionEvent).workflowId
```

Comment:
I suspected you might bring this up :-).

I tried several approaches. I think the only real way to get solid type support is to use inline functions in the machine config. That didn't fully solve the problem and it's specifically recommend against by the documentation. So my compromise was to at least narrow things down by individually defining the events I was casting to so that in the context of the action we know what the shape of the data payload should look like.

I had a similar problem with the `services` below here. I could not find a way with the current type definitions to make the return types of the services strongly types. At best, they will be `Promise<any>`. That means that the implementations that I provide when I interpret this machine in my react hook have no way of being checked that they are returning a value with the correct shape, property names, etc.

The good news is that when the actions and services _don't_ do what is expected, the tests will blow up. I spent a fair chunk of time getting tests working again, and some of that was finding places where actions/services were returning or expecting the wrong type. So I think for now, tests and runtime checking will be the best we can do.

I have two additional enhancements in mind for a follow-up PR:
1. Look into model-based testing for the machine. This should ensure that all machine states are reachable, that the rendered components look correct when we reach each state, and should fully exercise all of the actions/services. That should be a good signal for when things change and our typing isn't able to warn us at compile time.
2. Some people in the community have proposed workarounds involving runtime checks that are type guards. Basically, that we should be able to test the payload sent to the action and so a `is SelectWorkflowVersionEvent`. That won't help catch errors at compile-time though.

Reply:
I guess I really am that predictable that I'd bring up types in a PR 😆 

Sounds good, I certainly have spent less time with xstate than you have, so thanks for the thoughtful comments and I'm glad we have validation here in the form of tests!